keps/sig-api-machinery/NNNN-built-in-default/README.md

```diff
@@ -0,0 +1,302 @@
+based validation, validation is entirely out-of-scope for this project.
+
+This proposal also doesn't intend to remove the existing defaulting
+mechanism since some defaulting still requires complex logic that can't
+be defined declaratively. We could replace some defaulting that doesn't
+require to be done as code though.
+
+## Proposal
+
+Since the goal of this proposal is to increase consistency with the CRD
+defaulting, we propose to use the same mechanism (both in OpenAPI and
+similar to what kubebuilder does). This defaulting will happen BEFORE
+the existing defaulting functions are called (existing functions can be
+updated to panic so that we can see if defaults are properly applied).
+
+One very important nuance with CRDs is that CRDs are deserialized in
+unstructured objects that can distinguish between specified and
+unspecified fields. That is not the case for built-in types that are
+directly deserialized into Go-struct, where the unspecified fields are
+zero-valued. It's important to understand that only zero-value
+omit-empty fields will be defaulted using that mechanism.
+
+Complex defaulting will not be supported as a declarative marker and
+will continue being supported through the existing defaulting functions.
+
+### Risks and Mitigations
+
+The main risk is a bug that would cause a default to not be set as
+expected, either because the default would be missing or not set as
+expected. This would result in possible conversion failures, invalid
+objects being stored in the database or being served to users.
+
+## Design Details
+
+There are multiple challenges to running defaults declaratively and
+changing the current mechanisms:
+
+- We don't use OpenAPI to drive validation nor defaulting
+currently. There is no routing of the OpenAPI to most operations in
+kubernetes (the only thing right now is server-side apply).
+- OpenAPI fields don't strictly map to Go types fields
+(e.g. capitalization, embedding)
+- Reflection can't access comments, only structTags (which is why we have
+both structTags AND comments for SMP, but that's ugly and suboptimal).
+
+One possible way to address that second point is to serialize the object
+back to json, apply the default, and re-deserialize back into a
+go-struct, but that's memory intensive and Kubernetes doesn't need that.
+
+We're proposing to use the same mechanisms that we've put in place for
+server-side apply, implemented in
+"sigs.k8s.io/structured-merge-diff". We currently have a way for
+server-side apply to convert Go types into "json"-like values without
+serializing/deserializing and allocating memory by using reflection. We
+also have mechanisms in place to parse, process and use the OpenAPI.
+
+In other words, the plan would be the following:
+1. Update kube-openapi to parse the `+default` tag and include it in the
+OpenAPI definition
+2. Add additional "default" field to sigs.k8s.io/structured-merge-diff
+internal schema, and parse that field from the OpenAPI.
```

Comment:
are we "early enough" just after unmarshaling *before* conversion with SMD? Defaults apply to the version read from etcd and to the version sent by the client.

And this is important: objects from etcd are default. Afaik object from etcd don't go through SMD today.

Reply:
I'm actually changing the design to use generated code instead, but yes, we already are as early as possible with server-side apply so that shouldn't be a problem.

Reply:
The second point too? Defaulting on read?